doc/locking.md

```diff
@@ -0,0 +1,112 @@
+# Locking in Picolibc
+
+By default, picolibc is built with threading support, including
+various locks, enabled. However, picolibc has no scheduling support to
+control thread execution. That means it can't provide real
+implementations of the underlying locking primitives. Instead, it
+provides dummy functions sufficient for picolibc to work in a single
+threaded environment.
+
+If you are using picolibc in a re-entrant environment, like an RTOS
+with threading support, you will need to provide an implementation of
+this API on top of the scheduling primitives provided by the
+environment.
+
+## Where Picolibc uses locking
+
+Picolibc has a single global lock for APIs that share global
+data. That includes:
+
+ * malloc family
+ * onexit/atexit
+ * arc4random
+ * functions using timezones (localtime, et al)
+ * legacy stdio struct reent globals
+
+Tinystdio (the default stdio) uses per-file locks for the buffered
+POSIX file backend, but it doesn't require any locks for the bulk of
+the implementation. It uses atomic exchanges to handle the one
+reentrancy issue related to ungetc/getc.
+
+The legacy stdio implementation is full of locking, and has per-file
+locks for every operation.
+
+## Configuration options controlling locking
+
+There are two configuration options related to locking:
+
+ * newlib-retargetable-locking. When 'true', locking operations are
+   enabled and performed by the retargetable locking API described below.
+
+ * newlib-multithread. When 'true', the library is built with
+   multithreading support enabled and uses the locking operations.
+
+Picolibc inherits these options from newlib, and they're so
+interrelated as to make them effectively co-dependent, so users must
+set them to the same value.
+
+## Retargetable locking API
+
+When newlib-multithread and newlib-retargetable-locking are enabled
```

Comment:
There is "enabled enabled" at the beginning (duplicated). Let me have a go at this:

When newlib-multithread and newlib-retargetable-locking are enabled, Picolibc uses the following locking API. Picolibc provides stubs for all of these functions that do not actually perform any locking, so that a single-threaded application can still use a Picolibc compiled with locking enabled. An application needing locking must provide a real implementation of the locking API.

Reply:
That's way better.